### ISSUE TITLE
Additional customizations for radar charts

### ISSUE THREAD

**Issue by SommerEngineering** [swharden/ScottPlot#628]
Dear team,

Thank you very much for this great library 🙂

I would like to see the possible font options (e.g. size, bold, font) for the radar charts as well. As far as I can see (cf. [line 121 in RadarPlot.cs](https://github.com/swharden/ScottPlot/blob/63cc24c05912db00b1ea42b856869fe76ef923ae/src/ScottPlot/Plottable/RadarPlot.cs#L121)) the font cannot be configured right now. I would like to format the font for the axis and the category names differently. For me, this would be a valuable addition.

Also, the different axis of the radar seems to be automatic. I would need the opposite: Manually fixed axis. Let's say something like this:

```
var allValues = new [,] { { 5, 357, 0.244 } };
var categoryNames = new[] { "A", "B", "C" };
var fillColors = new[] {Color.FromArgb(128, Color.CornflowerBlue)};

plt.PlotRadar(values: allValues, categoryNames: categoryNames, categoryFontSize: 26, fillColors: fillColors);

plt.RadarAxis(axisIndex: 1, min: 1, max: 7, fontSize: 21, bold: true);
plt.RadarAxis(axisIndex: 2, min: 10, max: 750, fontSize: 21, bold: true);
plt.RadarAxis(axisIndex: 3, min: -1, max: 1, fontSize: 21, bold: true);
```

**Comment by swharden** [comment]
Hey @SommerEngineering, thanks for the suggestions!

Radar plots were first discussed in #428 and #430, and I'll tag @Benny121221 in case they want to comment.

#### Font Customization
I'll add some public fields to [RadarPlot.cs](https://github.com/swharden/ScottPlot/blob/master/src/ScottPlot/Plottable/RadarPlot.cs) so users can tweak the font. It'll be in ScottPlot 4.1 which might not get released for a little while though, so if you want an immediate solution you can download the ScottPlot 4.0 source code (it has its own branch) and make this modification yourself.

#### Radar Axis
I don't think I fully understand your goal here... Is your goal to be able to individually customize font options for each "arm" of the radar plot? Does your question somehow relate to the coordinate system in 2D space?

**Comment by Benny121221** [comment]
## Fonts

On the topic of font customization, it might be more convenient to have in the plot settings properties like `axisFont`, `tickFont`, `legendFont` and `plotFont`, where `plotFont` is the font that plottables use, be it a Radar, Pie, or Bar plot. If you wanted to simplify it you could have just `frameFont` and `plotFont`, where `frameFont` is the font for the axis, ticks, legend, etc. I think this is likely more convenient for the user, however, it has the downside that all plottables will have the same font. I'm unsure how much this downside matters to you. It's worth noting that this is not mutually exclusive with taking a custom font per plottable, that could override a default `plotFont`.

## Axes
As for the radar axis, do you mean each leg being independently scaled? For example, if I add points to this demo it looks very bad, because both drivers receive many more points than they do anything else:

Without WDC Points:
![image](https://user-images.githubusercontent.com/8635304/100271711-2cf5ed80-2f17-11eb-9e2b-8f39122b3470.png)

With WDC Points:
![image](https://user-images.githubusercontent.com/8635304/100271628-089a1100-2f17-11eb-92ff-15fe8f7cbe3d.png)

In principle, you could rescale these axes so that points was on a scale from 0-256 (Vettel scored 256 WDC points in his inaugural title year), DNFS from 0-3, etc. However, I feel this would be very cluttered unless you were to scrap the axis labels, as otherwise you would need one on each leg of the plot.

So, if I understand you correctly, you want each leg to support a separate axis. In practice, this would almost certainly be implemented as multiple scaling factors (Radar plots and Pie plots are drawn with a radius of 1 in terms of whichever axis has smaller units, they have only a loose relationship with the coordinate system). I feel this solution would get messy unless one were to remove the labels. And if you remove the labels it becomes practically equivalent to rescaling all the axes onto the interval [0,1]. This becomes equivalent to plotting the fraction of say, Pole Positions that Alonso scored vs Vettel.

![image](https://user-images.githubusercontent.com/8635304/100271562-e99b7f00-2f16-11eb-99bd-ef60998a7cae.png)

The only difference is that this still has axis labels as it's an axis from 0 to 1. I'm actually moderately surprised I didn't add an option to remove labels in the original PR.

**Comment by Benny121221** [comment]
I wonder if such rescaling would be wise to include in `ScottPlot.Tools` or `ScottPlot.Statistics`? There are already a few implementations of min-max normalization (and rescaling, which is a generalization with min := 0), for example in `PlottableHeatmap`, `PlottableRadar`, etc.

I'm aware that this isn't a statistics library, but it would probably be a good idea, it saves us work if we ever create `PlottableHeatmappedPieRadar` or something wonky like that, and it saves the user some time from either implementing their own or having to read up docs on another library. If someone comes around with 100 million points it might make sense for them to implement it in C/C++, perhaps with some fancy CUDA/OpenCL acceleration (or use a stats library that does), but those people know who they are.

**Comment by swharden** [comment]
Hey @Benny121221, thanks for your thorough response! I'll hold-off on a details answer until we hear back from @SommerEngineering regarding the goal for the spider plot.

### Global Font Settings

Your suggestions are interesting ones, but I think my current lean is to keep font settings simple, separate, and contained in their respective renderables. As I move into ScottPlot 4.1 I'm trying to reduce the aggregation of global state and do a better job drawing clean lines around modules and reducing coupling between them. For now boring public fields are probably the way to go.

The "enterprise solution" is probably to have an `IHasFont` with fields like `FontName`, `FontSize`, and `FontScale`, and to have renderables that use fonts implement it. Then a `Plot.SetFont()` method could iterate through all the renderables in the plot that use fonts and tweak their style. That's a lot of added fanfare through, and _another_ interface new contributors have to keep a look out for when making new plottables... so I think it's probably not worth the added complexity. Having the user tweak axis label, axis tick, and plottable font settings according to their taste isn't too big of an ask for now.

### Radar Math

My lean on where to put that new logic would be inside the radar class since it's the only class that uses it at this time. I agree that future plot types may use that math, but those plot types don't exist and my lean is to only extract that logic out into another class when that time comes. Otherwise it's another module to maintain (and maintain tests for?), so my lean is to hide it in the radar class for now. I think this falls under YAGNI?

Thanks again for your input here @Benny121221!

**Comment by SommerEngineering** [comment]
Thanks @swharden and @Benny121221 for your quick responses 🙂 

# Fonts
For me it is only important that I can somehow adjust the font. I cannot give an estimation, which implementation is best. You have the overall view.

From a user's point of view, I see it like this: If there was only one global setting, but I needed different fonts for different charts, I would create new instances of the plot class each time. So I see this very pragmatically.

Thanks @swharden for your hint to checkout the 4.0 branch to apply a temporary change.

# Axes

## My Use Case
Thanks @Benny121221 for your input. Thus I realized that my description was not sufficient. You are right: The axes must all have the same length. My point is that the axes are labeled differently. Let me give an example.

Let's say I have to analyze and compare different technologies. That's what the Technology Readiness Level (TRL) from NASA is for. The scale goes from 1 to 9. In addition to the TRL, three other metrics (let's call them just A, B, and C) with a scale of 1-5 are required.

So I need a total of 4 arms (axes). All arms are of equal length in the diagram, say 200 pixels as an example. However, one arm is labeled 1-9 and the others 1-5.

It is also important: The labeling of the arms must be static, no matter what actual values are displayed. So the TRL arm must always show 1-9, even if my technologies only use TRL 1, 3 and 4. Why? The viewer should see that the TRL ranges from 1-9, so there is room for further developments.

In my PhD thesis I use radar charts where one arm is labeled e.g. 1-100000 and in the same chart another arm 1-10. Here, as well, the label should always show the full range, so that you can see what the theoretical maximum is. (For my thesis I created the charts in the R environment)

## Implementation
Yes, I agree with you: the implementation is not easy. It is, so to speak, a decoupling between the actual length of the arms and their labeling.

Say all arms (and their values) are normalized to the interval [0,1]. If the user now wishes 1-9 for an arm, the code would have to map the desired interval (axis labeling) to the actual interval. So for TRL 1-9 we have 9 steps (the user could set the step size explicitly to 1), converted to [0,1] each step would be 0,125:

Desired label | Position on actual interval
---------------|----------------------------
TRL 1 | 0,000
TRL 2 | 0,125
TRL 3 | 0,250
TRL 4 | 0,375
TRL 5 | 0,500
TRL 6 | 0,625
TRL 7 | 0,750
TRL 8 | 0,875
TRL 9 | 1,000

So far, so simple. What if a user now needs a scale from 1-100000 for an arm? This could then be represented logarithmically (the user should configure this).

Desired label | Position on actual interval
---------------|----------------------------
10^0 | 0,0
10^1 | 0,2
10^2 | 0,4
10^3 | 0,6
10^4 | 0,8
10^5 | 1,0

I hope that my description now clearly explains the use case. Unfortunately, I don't have time to help with the implementation at the moment.

**Comment by SommerEngineering** [comment]
Note: here is an example for Gnuplot, how to archive my use case: https://stackoverflow.com/a/56551726/2258393

**Comment by Benny121221** [comment]
Intriguing, in principle, the normalization could be done by adding a parameter like `INormalizationStrategy[] normalizationStrategy`, however I think the following is more elegant:

- Add option for multiple axis labels on a Radar Plot
    - This means each axis should be rescaled separately, with no option for fancy rescaling (except perhaps a non-zero minimum)
    - If the user wants logarithmic rescaling they should take the log before passing it into the radar plot
          - Conveniently there is `ScottPlot.Tools.Log10` for this purpose, and once normalized onto [a,1] all bases are equivalent ∀ a

While it wouldn't be much of an addition to add an option for log data it would open the door to accepting sigmoid-ed data and standardized and cosine-ized and bucketed and data injectively mapped to an analytic continuation of a planar cross section of an average tapir's mouth.

All of that (except perhaps the last one) would be useful, but not realistically implemented on every plottable. This is why the demos on log and polar plots use `Log10` and `ConvertPolarCoordinates` from `ScottPlot.Tools` instead of a sneaky new parameter on every plottable for which it would feasibly make sense.

**Comment by swharden** [comment]
I added customizable font support to radar plots in 4.1. I didn't go back and add it to 4.0, but 825f734debc36f6d367a524d3c14aced16e7b9ed indicates the code that requires change.

I created an issue (#631) to remind me to review code for every plottable and ensure they all have customizable font options.

```cs
var radar = plt.PlotRadar(values, categories, groups);
radar.Font.Name = "Comic Sans MS";
radar.Font.Size = 16;
radar.Font.Color = Color.Magenta;
```

![image](https://user-images.githubusercontent.com/4165489/100376326-58431000-2fdd-11eb-86cb-2ac80c70c19f.png)

The remaining question is if/how to further customize the radar plot as described above. I'm working pretty hard every day on #578 so I don't plan on implementing this myself. I see @SommerEngineering said "I don't have time to help with the implementation at the moment", so it's probably up to @Benny121221 whether this is something you're interested in pursuing at this time. 

If so, working on the master branch probably makes the most sense. I'm actively working on the axis (and multi-axis) system, but as long as your work is confined to the RadarPlot class and we won't get merge conflicts.

If not, I can close this issue (and add it to #412?) and we can pick it up if/when a better time comes.

**Comment by Benny121221** [comment]
> so it's probably up to @Benny121221 whether this is something you're interested in pursuing at this time.

I got some midterms to study for (despite it being well past the middle of term....) so I don't see myself working on this before the weekend. However, I would be happy to implement it.

**Comment by swharden** [comment]
> I would be happy to implement it

That's great to hear, thanks @Benny121221! Take your time (doing well on midterms is certainly more important!), and if you change your mind, get busy, or there isn't movement on this issue for a while I'll close it and we can pick it up again whenever a better time arises 👍

**Comment by swharden** [comment]
@Benny121221 thanks for adding those additional customizations! I just merged #634 but wanted to echo your idea from that PR:

> Makes me wonder if there should be an option to remove the circular rings? Or replace them with polygonal ones.

I do like the look of the straight lines... it definite looks more like a spider web! I also recognize that many other charting libraries display radar plots like that, including the gnuplot example @SommerEngineering pointed to in an earlier comment 🤔

That's definitely icing on the cake though, so we can close this issue now if you think it's in a good spot, or I can leave it open if you wanted to have a go at supporting straight line line (rather than circle) axes in radar plots

**Comment by swharden** [comment]
This looks fantastic @Benny121221, way to go!

`RadarAxis.Circle` | `RadarAxis.Polygon`
---|---
![image](https://user-images.githubusercontent.com/4165489/100530732-8124fa00-31c3-11eb-8020-7609952265ce.png)|![image](https://user-images.githubusercontent.com/4165489/100530733-87b37180-31c3-11eb-912d-0a4276b2e147.png)

I think we addressed all of @SommerEngineering's original suggestions, but if not feel free to open this issue back up 👍 

I hope to get a ScottPlot 4.1 pre-release out on NuGet tomorrow 🤞

**Comment by SommerEngineering** [comment]
Wow, many thanks for the fast and professional implementation :slightly_smiling_face: . It has become awesome. Thanks also for the example in PR :+1: .

The last thing that was missing was that I could define the maximum of the axes myself, instead of deriving it from the given values. I created a PR for this.

**Comment by swharden** [comment]
> The last thing that was missing was that I could define the maximum of the axes myself, instead of deriving it from the given values. I created a PR for this.

Thanks @SommerEngineering! I merged #636 then refined the cookbook example names/descriptions a bit. Thanks for the contribution! 👍
